src/mtype.d

```diff
@@ -10244,15 +10264,95 @@ extern (C++) final class Parameter : RootObject
+         *   stc = storage class of parameter
+         */
+        static uint buildSR(bool returnByRef, StorageClass stc) pure nothrow @nogc @safe
+        {
+            uint result;
+            final switch (stc & (STCref | STCscope | STCreturn))
+            {
+                case 0:                    result = SR.None;        break;
+                case STCref:               result = SR.Ref;         break;
+                case STCscope:             result = SR.Scope;       break;
+                case STCreturn | STCref:   result = SR.ReturnRef;   break;
+                case STCreturn | STCscope: result = SR.ReturnScope; break;
+                case STCref    | STCscope: result = SR.RefScope;    break;
+                case STCreturn | STCref | STCscope:
+                    result = returnByRef ? SR.ReturnRef_Scope : SR.Ref_ReturnScope;
+                    break;
+            }
+            return result;
+        }
+
+        /* result is true if the 'from' can be used as a 'to'
+         */
+
+        if ((from ^ to) & STCref)               // differing in 'ref' means no covariance
+            return false;
 
-                // We can add scope, but not subtract it
-                (!(this.storageClass & STCscope) && (p.storageClass & STCscope)) ||
+        return covariant[buildSR(returnByRef, from)][buildSR(returnByRef, to)];
+    }
+
+    /* Classification of 'scope-return-ref' possibilities
+     */
+    enum SR
+    {
+        None,
+        Scope,
+        ReturnScope,
+        Ref,
+        ReturnRef,
+        RefScope,
+        ReturnRef_Scope,
+        Ref_ReturnScope,
+    }
+
+    extern (D) static immutable bool[SR.max + 1][SR.max + 1] covariant;
+
+    static this()
+    {
+        /* Initialize covariant[][] with this:
+
+             From\To           n   rs  s
+             None              X
+             ReturnScope       X   X
+             Scope             X   X   X
+
+             From\To           r   rr  rs  rr-s r-rs
+             Ref               X   X
+             ReturnRef             X
+             RefScope          X   X   X   X    X
+             ReturnRef-Scope       X       X
+             Ref-ReturnScope   X   X            X
+        */
+        foreach (i; 0 .. SR.max + 1)
+        {
+            covariant[i][i] = true;
+            covariant[SR.RefScope][i] = true;
+        }
+        covariant[SR.ReturnScope][SR.None]        = true;
+        covariant[SR.Scope      ][SR.None]        = true;
+        covariant[SR.Scope      ][SR.ReturnScope] = true;
 
-                // We can subtract return, but not add it
-                ((this.storageClass & STCreturn) && !(p.storageClass & STCreturn)));
+        covariant[SR.Ref            ][SR.ReturnRef] = true;
+        covariant[SR.ReturnRef_Scope][SR.ReturnRef] = true;
+        covariant[SR.Ref_ReturnScope][SR.Ref      ] = true;
+        covariant[SR.Ref_ReturnScope][SR.ReturnRef] = true;
```

Comment:
Thanks. I thought the error came from running one of the test programs, not the compiler itself. But there's nothing platform specific there - why is the error happening only one some platforms?